### PR TITLE
Add .travis.yml and build scripts for Linux and Mac OS X

### PR DESCRIPTION
**Pull request by saudet**
## What changes were proposed in this pull request?

Let Travis CI make builds and deploy SNAPSHOT artifacts to Sonatype OSS automatically for CPU and CUDA on Linux and Mac OS X.

## How was this patch tested?

Builds for Linux and Mac OS X pass on Travis CI.

### CODE REVIEW

File: ci/build-macosx-x86_64.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+fi
+
+brew update > /dev/null
+brew install gcc || true
+brew link --overwrite gcc
+
+if [[ $CUDA == "8.0" ]]; then
+    curl -L https://developer.nvidia.com/compute/cuda/8.0/Prod2/local_installers/cuda_8.0.61_mac-dmg -o $HOME/cuda.dmg
+else
+    curl -L https://developer.nvidia.com/compute/cuda/9.0/Prod/local_installers/cuda_9.0.176_mac-dmg -o $HOME/cuda.dmg
```
Comment (agibsonccc) [inline ci/build-macosx-x86_64.sh]: We might want to host these ourselves.

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+dist: trusty
```
Comment (agibsonccc) [inline .travis.yml]: Isn't fedora required? I noticed that the build down below uses RPMs.

File: ci/build-linux-x86_64.sh [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    MAVEN_PHASE="install"
+fi
+
+if ! git -C $TRAVIS_BUILD_DIR/.. clone https://github.com/deeplearning4j/libnd4j/ --depth=50 --branch=$TRAVIS_BRANCH; then
+     git -C $TRAVIS_BUILD_DIR/.. clone https://github.com/deeplearning4j/libnd4j/ --depth=50
+fi
+
+docker run -ti -e SONATYPE_USERNAME -e SONATYPE_PASSWORD -v $HOME/.m2:/root/.m2 -v $TRAVIS_BUILD_DIR/..:/build \
+    nvidia/cuda:$CUDA-cudnn$CUDNN-devel-centos6 /bin/bash -evxc "\
+        yum -y install centos-release-scl-rh epel-release; \
```
Comment (agibsonccc) [inline ci/build-linux-x86_64.sh]: Flagging for above, it seems like we're using ubuntu but we're using RHEL commands down here.

File: .travis.yml (hunk 2 above)
Comment (saudet) [reply .travis.yml]: Travis only supports Trusty, that's why we need to use Docker.

File: ci/build-linux-x86_64.sh (hunk 3 above)
Comment (saudet) [reply ci/build-linux-x86_64.sh]: Yes, those commanes are running it Docker.

File: ci/build-macosx-x86_64.sh (hunk 1 above)
Comment (saudet) [reply ci/build-macosx-x86_64.sh]: We'll see, I guess? Going through Microsoft, Amazon, or NVIDIA, not sure there is a big difference...

**Comment by saudet** [comment]
I got builds working for Android as well!

@agibsonccc So, anything that we need to change before we can merge this?

**Pull request merged**